### ISSUE TITLE
Add an option to append genres

### ISSUE THREAD

**Issue by arsaboo** [beetbox/beets#4750]
The current implementation of `lastgenre` ignores any existing genres that a user may have tagged and completely overwrites the genres returned by lastgenre. It would be nice to have the option to append the new genres returned by LastGenre to those that are already available. 

### Proposed solution

Extend the plugin to add an append option.

**Comment by sampsyo** [comment]
Sounds like a reasonable addition!

**Comment by JOJ0** [comment]
Maybe that is a start for fixing this? https://github.com/beetbox/beets/commit/03848c6bb948dbc3cbd9e9f82009c1df1366ed7f

**Comment by arsaboo** [comment]
> Maybe that is a start for fixing this? [03848c6](https://github.com/beetbox/beets/commit/03848c6bb948dbc3cbd9e9f82009c1df1366ed7f)

Does it take care of the fact that the pre-existing genre may be duplicated, i.e., the pre-existing genre may also be included in the new genre tags?

**Comment by JOJ0** [comment]
Nope doesnt yet. It might be _a start_ you/we can work from. The main reason I started tinkering with lastgenre was that it _always_ overwrites when more than one genre (comma separated) was existing in the genre field/tag because this check doesnt handle it.

**Comment by arsaboo** [comment]
@JOJ0 I have also been working on something along those lines. I have a working prototype - https://github.com/arsaboo/beets/tree/append_genre

Ignore the `all_genres` part for now, which I only added so that I can see what genres are being obtained (so that I can clean them up later). It basically, reads the original genre tags (even if they are more than one) and append the new tags. Tags are deduplicated. Let me know what you think.

**Comment by JOJ0** [comment]
Cant see much w the link. on mobile. travelling right now. maybe send commit link.

but probably we should combine our efforts at some point and things/behaviour could be made configurable.

my approach tries to only allow what genres.txt defines and also later (not in that commit) I added cleanup via aliases and regex replacements. For example genre "drum n bass" (and many other spellings) will be accepted either as an existing tag or when received from last.fm but will be "normalized" to "Drum And Bass" during all checks and also when finally being saved to library/tag.

**Comment by JOJ0** [comment]
More details on what I tinkered so far: https://github.com/beetbox/beets/discussions/4765#discussioncomment-6060849

**Comment by arsaboo** [comment]
>  For example genre "drum n bass" (and many other spellings) will be accepted either as an existing tag or when received from last.fm but will be "normalized" to "Drum And Bass" during all checks and also when finally being saved to library/tag.
I don't think we need regex for this. We can use canonicalization to achieve the same. I agree, we should combine our efforts. Maybe I will create a draft PR for you to look at and try things out.

**Comment by JOJ0** [comment]
Maybe we are misunderstanding each orher? I dont use cannonicalization to be honest. Maybe dnb was a bad example. Another one: "2 step garage" should be just "2-step"

whatlastgenre has 2 approaches to solve this: a simple alias table (fast), regex replacements (expensive)

Anyway, yeah, I'll try to submit the things I borrowed from whatlastgenre when Im back home around in 2 weeks.

Send me links to your commits, I'm still curious :-) Thanks!

**Comment by arsaboo** [comment]
You should still be able to use canonicalization for that. All you need in your genre-trees is:
```yaml
- 2 step
    -  2 step garage
```
I have yet to come across an example that I cannot manage using canonicalization....try it out. 

All this assumes we know what genres (and their variants) can be used in several tag sources. This is why I added the logging option that collects all the genre tags. 

I have created a draft PR #https://github.com/beetbox/beets/pull/4811 for you to look at. It has most of the things that we discussed. Let me know what you think and let us get this done. 

Another option (maybe in the long run) is to create a generic `genre` plugin that can take any genre source (something like wlg).

**Comment by JOJ0** [comment]
Sure, but fixing spelling is not the purpose of canonicalization. 

Thanks for the draft, I'll take a look when I'm back home from holiday.

**Comment by arsaboo** [comment]
> Sure, but fixing spelling is not the purpose of canonicalization.

Why not? I like to think of canonicalization as just a way to organize my genres. In any case, it doesn't hurt to provide a regex option if users prefer to use that.
